dbms/src/Functions/greatCircleDistance.cpp

```diff
@@ -14,70 +14,80 @@
- *  Latitude must be in [-90, 90], longitude must be [-180, 180].
- *  Original code of this implementation of this function is here https://github.com/sphinxsearch/sphinx/blob/409f2c2b5b2ff70b04e38f92b6b1a890326bad65/src/sphinxexpr.cpp#L3825.
- *  Andrey Aksenov, the author of original code, permitted to use this code in ClickHouse under the Apache 2.0 license.
- *  Presentation about this code from Highload++ Siberia 2019 is here https://github.com/ClickHouse/ClickHouse/files/3324740/1_._._GEODIST_._.pdf
- *  The main idea of this implementation is optimisations based on Taylor series, trigonometric identity and calculated constants once for cosine, arcsine(sqrt) and look up table.
- */
+/** Calculates the distance between two geographical locations.
+  * There are two variants:
+  * greatCircleDistance: calculates the distance on a sphere: https://en.wikipedia.org/wiki/Great-circle_distance
+  * geoDistance: calculates the distance on WGS-84 ellipsoid.
+  *
+  * The function calculates distance in meters between two points on Earth specified by longitude and latitude in degrees.
+  *
+  * Latitude must be in [-90, 90], longitude must be [-180, 180].
+  *
+  * Original code of this implementation of this function is here:
+  * https://github.com/sphinxsearch/sphinx/blob/409f2c2b5b2ff70b04e38f92b6b1a890326bad65/src/sphinxexpr.cpp#L3825.
+  * Andrey Aksenov, the author of original code, permitted to use this code in ClickHouse under the Apache 2.0 license.
+  * Presentation about this code from Highload++ Siberia 2019 is here https://github.com/ClickHouse/ClickHouse/files/3324740/1_._._GEODIST_._.pdf
+  * The main idea of this implementation is optimisations based on Taylor series, trigonometric identity
+  *  and calculated constants once for cosine, arcsine(sqrt) and look up table.
+  */
 
 namespace
 {
 
 constexpr double PI = 3.14159265358979323846;
-constexpr float TO_RADF = static_cast<float>(PI / 180.0);
-constexpr float TO_RADF2 = static_cast<float>(PI / 360.0);
+constexpr float RAD_IN_DEG = static_cast<float>(PI / 180.0);
+constexpr float RAD_IN_DEG_HALF = static_cast<float>(PI / 360.0);
+
+constexpr size_t COS_LUT_SIZE = 1024; // maxerr 0.00063%
+constexpr size_t ASIN_SQRT_LUT_SIZE = 512;
+constexpr size_t METRIC_LUT_SIZE = 1024;
+
+/// Earth mean diameter in meters, https://en.wikipedia.org/wiki/Earth
+constexpr float EARTH_DIAMETER = 2 * 6371000;
+
+float cos_lut[COS_LUT_SIZE + 1];       /// cos(x) table
+float asin_sqrt_lut[ASIN_SQRT_LUT_SIZE + 1]; /// asin(sqrt(x)) * earth_diameter table
 
-constexpr size_t GEODIST_TABLE_COS = 1024; // maxerr 0.00063%
-constexpr size_t GEODIST_TABLE_ASIN = 512;
-constexpr size_t GEODIST_TABLE_K = 1024;
+float sphere_metric_lut[METRIC_LUT_SIZE + 1];    /// sphere metric: the distance for one degree across longitude depending on latitude
+float wgs84_metric_lut[2 * (METRIC_LUT_SIZE + 1)];  /// ellipsoid metric: the distance across one degree latitude/longitude depending on latitude
```

Comment:
If I understand correctly compiler would work with two-dimensional like with one-dimensional, so we can have an impact on performance only by access order (by choice what are rows and what are columns). But two-dimensional array looks simplier to read.

Reply:
Two dimensional array will be placed in one contiguous place in memory.
But access to that array will be represented as double pointer redirection.

Reply:
Actually you are right, it will be the same :)
https://godbolt.org/z/fwuQ8d